File: src/activation_circuit.cpp
#include "activation_circuit.hpp"

#include <algorithm>
#include <cmath>
#include <cctype>
#include <string>
#include <string_view>

namespace volt {

namespace {

std::string lower(std::string_view s) {
    std::string out(s);
    for (char& c : out) {
        c = static_cast<char>(std::tolower(static_cast<unsigned char>(c)));
    }
    return out;
}

float sigmoid_circuit(float I_in, const Config& cfg) {
    const float mid = cfg.I_min + 0.5f * cfg.I_range;
    const float scale = cfg.I_range * 0.25f + 1e-30f;
    const float x = (I_in - mid) / scale * cfg.circuit_steepness;
    const float s = 1.0f / (1.0f + std::exp(-x));
    return cfg.I_min + cfg.I_range * s;
}

double sigmoid_circuit(double I_in, const Config& cfg) {
    const double mid = static_cast<double>(cfg.I_min) + 0.5 * static_cast<double>(cfg.I_range);
    const double scale = static_cast<double>(cfg.I_range) * 0.25 + 1e-30;
    const double x =
        (I_in - mid) / scale * static_cast<double>(cfg.circuit_steepness);
    const double s = 1.0 / (1.0 + std::exp(-x));
    return static_cast<double>(cfg.I_min) + static_cast<double>(cfg.I_range) * s;
}

}  // namespace

float circuit_transfer(float I_in, CircuitModel model, const Config& cfg) {
    switch (model) {
        case CircuitModel::PassThrough:
            return I_in;
        case CircuitModel::DiodeRectifier:
            return std::max(0.0f, I_in - cfg.circuit_i_threshold);
        case CircuitModel::TunableSigmoid:
            return sigmoid_circuit(I_in, cfg);
        default:
            return I_in;
    }
}

double circuit_transfer(double I_in, CircuitModel model, const Config& cfg) {
    switch (model) {
        case CircuitModel::PassThrough:
            return I_in;
        case CircuitModel::DiodeRectifier:
            return std::max(0.0, I_in - static_cast<double>(cfg.circuit_i_threshold));
        case CircuitModel::TunableSigmoid:
            return sigmoid_circuit(I_in, cfg);
        default:
            return I_in;
    }
}

bool parse_circuit_model(std::string_view name, CircuitModel& out) {
    const std::string k = lower(name);
    if (k == "pass_through" || k == "passthrough" || k == "identity") {
        out = CircuitModel::PassThrough;
        return true;
    }
    if (k == "diode_rectifier" || k == "diode" || k == "relu") {
        out = CircuitModel::DiodeRectifier;
        return true;
    }
    if (k == "tunable_sigmoid" || k == "sigmoid") {
        out = CircuitModel::TunableSigmoid;
        return true;
    }
    return false;
}

const char* circuit_model_name(CircuitModel model) {
    switch (model) {
        case CircuitModel::PassThrough:
            return "pass_through";
        case CircuitModel::DiodeRectifier:
            return "diode_rectifier";
        case CircuitModel::TunableSigmoid:
            return "tunable_sigmoid";
        default:
            return "pass_through";
    }
}

}  // namespace volt

File: crates/volt-core/src/activation_circuit.rs
use crate::config::{CircuitModel, Config};

fn sigmoid_circuit_f32(i_in: f32, cfg: &Config) -> f32 {
    let mid = cfg.i_min + 0.5 * cfg.i_range;
    let scale = cfg.i_range * 0.25 + 1e-30;
    let x = (i_in - mid) / scale * cfg.circuit_steepness;
    let s = 1.0 / (1.0 + (-x).exp());
    cfg.i_min + cfg.i_range * s
}

fn sigmoid_circuit_f64(i_in: f64, cfg: &Config) -> f64 {
    let mid = cfg.i_min as f64 + 0.5 * cfg.i_range as f64;
    let scale = cfg.i_range as f64 * 0.25 + 1e-30;
    let x = (i_in - mid) / scale * cfg.circuit_steepness as f64;
    let s = 1.0 / (1.0 + (-x).exp());
    cfg.i_min as f64 + cfg.i_range as f64 * s
}

pub fn circuit_transfer_f32(i_in: f32, model: CircuitModel, cfg: &Config) -> f32 {
    match model {
        CircuitModel::PassThrough => i_in,
        CircuitModel::DiodeRectifier => (i_in - cfg.circuit_i_threshold).max(0.0),
        CircuitModel::TunableSigmoid => sigmoid_circuit_f32(i_in, cfg),
    }
}

pub fn circuit_transfer_f64(i_in: f64, model: CircuitModel, cfg: &Config) -> f64 {
    match model {
        CircuitModel::PassThrough => i_in,
        CircuitModel::DiodeRectifier => {
            (i_in - cfg.circuit_i_threshold as f64).max(0.0)
        }
        CircuitModel::TunableSigmoid => sigmoid_circuit_f64(i_in, cfg),
    }
}

pub fn parse_circuit_model(name: &str) -> Option<CircuitModel> {
    match name.to_ascii_lowercase().as_str() {
        "pass_through" | "passthrough" | "identity" => Some(CircuitModel::PassThrough),
        "diode_rectifier" | "diode" | "relu" => Some(CircuitModel::DiodeRectifier),
        "tunable_sigmoid" | "sigmoid" => Some(CircuitModel::TunableSigmoid),
        _ => None,
    }
}

pub fn circuit_model_name(model: CircuitModel) -> &'static str {
    match model {
        CircuitModel::PassThrough => "pass_through",
        CircuitModel::DiodeRectifier => "diode_rectifier",
        CircuitModel::TunableSigmoid => "tunable_sigmoid",
    }
}
